tracer/src/Datadog.Trace/Telemetry/TelemetryDataBuilder.cs

```diff
@@ -0,0 +1,109 @@
+
+            if (configuration is not null)
+            {
+                Log.Debug("App initialized, sending app-started");
+                var payload = new AppStartedPayload
+                {
+                    Configuration = configuration,
+                    Dependencies = dependencies,
+                    Integrations = integrations,
+                };
+
+                return new[] { GetRequest(application, host, TelemetryRequestTypes.AppStarted, payload) };
+            }
+
+            if (dependencies is null && integrations is null)
+            {
+                if (sendHeartbeat)
+                {
+                    Log.Debug("No changes in telemetry, sending heartbeat");
+                    return new[] { GetRequest(application, host, TelemetryRequestTypes.AppHeartbeat, payload: null) };
+                }
+
+                Log.Debug("No changes in telemetry");
+                return Array.Empty<TelemetryData>();
+            }
+
+            if (dependencies is null)
+            {
+                Log.Debug("Integrations updated, sending app-integrations-change");
+                var payload = new AppIntegrationsChangedPayload { Integrations = integrations };
+
+                return new[] { GetRequest(application, host, TelemetryRequestTypes.AppIntegrationsChanged, payload), };
+            }
+
+            if (integrations is null)
+            {
+                Log.Debug("Dependencies updated, sending app-dependencies-loaded");
+                var payload = new AppDependenciesLoadedPayload { Dependencies = dependencies };
+
+                return new[] { GetRequest(application, host, TelemetryRequestTypes.AppDependenciesLoaded, payload), };
+            }
+
+            Log.Debug("Dependencies updated, sending app-dependencies-loaded");
+            Log.Debug("Integrations updated, sending app-integrations-change");
+            var depsPayload = new AppDependenciesLoadedPayload { Dependencies = dependencies };
+            var integrationsPayload = new AppIntegrationsChangedPayload { Integrations = integrations };
+
+            return new[]
+            {
+                GetRequest(application, host, TelemetryRequestTypes.AppDependenciesLoaded, depsPayload),
+                GetRequest(application, host, TelemetryRequestTypes.AppIntegrationsChanged, integrationsPayload),
+            };
+        }
+
+        public TelemetryData BuildAppClosingTelemetryData(ApplicationTelemetryData application, HostTelemetryData host)
+        {
+            return GetRequest(application, host, TelemetryRequestTypes.AppClosing, payload: null);
+        }
+
+        private TelemetryData GetRequest(
+            ApplicationTelemetryData application,
+            HostTelemetryData host,
+            string requestType,
+            IPayload payload)
+        {
+            var sequence = Interlocked.Increment(ref _sequence);
+
+            return new TelemetryData
+            {
+                SeqId = sequence,
```

Comment:
Does `api_version` also need to be sent as a top-level key (just like the request_type)? According to the Confluence page, it seems to be required at this moment

Reply:
Ugh silly me, it's a default value in `TelemetryData` so that's why it wasn't specified here. Sorry for the noise!